src/lib/block_producer/block_producer.ml

```diff
@@ -1197,24 +1199,31 @@ let run ~context:(module Context : CONTEXT) ~vrf_evaluator ~prover ~verifier
                     ~constants:consensus_constants ~consensus_state
                     ~local_state:consensus_local_state
                    = None ) ; *)
-            let next_vrf_check_now () = return (new_global_slot, i') in
+            let next_vrf_check_now () =
+              (* WARN: This recursive call will update [epoch_data_for_vrf] and
+                 in turn [new_epoch] and [new_global_slot]. Refactors aiming at
+                 removing this should consider structuring the code so updating
+                 on those data are included. *)
+              iteration_wrapped (new_global_slot, new_epoch)
```

Comment:
This is the fix.

Reply:
I don't understand why is this change needed.

It would just be the same to have `return (new_global_slot, new_epoch)`, it would be returned to `forever`, which would immediately call `iteration_wrapped` with it.